src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1299,6 +1299,34 @@ public static double Round(double value, int digits)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static double Round(double value, MidpointRounding mode)
         {
+            if (AdvSimd.Arm64.IsSupported)
+            {
+                // Inline the most common cases
+                if (mode == MidpointRounding.ToEven)
+                {
+                    return Round(value);
+                }
+                if (mode == MidpointRounding.AwayFromZero)
+                {
+                    return AdvSimd.Arm64.RoundAwayFromZero(Vector128.CreateScalarUnsafe(value)).ToScalar();
+                }
+            }
+            if (Sse41.IsSupported)
+            {
+                // Inline the most common cases
+                if (mode == MidpointRounding.ToEven)
+                {
+                    return Round(value);
+                }
+                if (mode == MidpointRounding.AwayFromZero)
+                {
+                    Vector128<double> valueVec = Vector128.CreateScalarUnsafe(value);
+                    Vector128<double> tmp = Sse2.And(valueVec, Vector128.Create(0x8000000000000000UL).AsDouble());
+                    tmp = Sse2.Or(tmp, Vector128.Create(0x3fdfffffffffffffUL).AsDouble());
+                    tmp = Sse2.Add(tmp, valueVec);
+                    return Sse41.RoundToZeroScalar(tmp).ToScalar();
```

Comment:
This has multiple "magic" variables and needs documentation explaining what its doing.

It's also not the implementation I would've expected here. I would've expected basically (which, iirc, can be optimized down to like 3 instructions):
```csharp
var tmp = value;
var result = RountTowardPositiveInfinity(Abs(value));
return CopySign(result, tmp);
```

Reply:
> which, iirc, can be optimized down to like 3/4 instructions

Can't fit your impl into 4, it's 6 for me:
```csharp
public static double RoundAWS_Tanner(double value)
{
    var tmp = Vector128.CreateScalarUnsafe(value);
    var abs = Sse2.And(tmp, Vector128.Create(0x7FFFFFFFFFFFFFFFUL).AsDouble());
    var result = Sse41.RoundToPositiveInfinityScalar(abs);
    // CopySign impl:
    var selector = Vector128.CreateScalarUnsafe(-0.0); // can't be contained since we use it twice
    return Sse2.Or(Sse2.And(result, selector), Sse2.AndNot(selector, tmp)).ToScalar();
}
```
```
; Method Program:RoundAWS_Tanner(double):double
G_M52853_IG01:
       vzeroupper 
						;; bbWeight=1    PerfScore 1.00

G_M52853_IG02:
       vmovsd   xmm1, qword ptr [reloc @RWD16]
       vandpd   xmm2, xmm0, xmmword ptr [reloc @RWD00]

       vroundsd xmm2, xmm2, xmm2, 2
       vandpd   xmm2, xmm2, xmm1
       vandnpd  xmm0, xmm1, xmm0
       vorpd    xmm0, xmm2, xmm0
						;; bbWeight=1    PerfScore 13.00

G_M52853_IG03:
       ret      
						;; bbWeight=1    PerfScore 1.00
RWD00  	dq	7FFFFFFFFFFFFFFFh, 7FFFFFFFFFFFFFFFh
RWD16  	dq	8000000000000000h	;           -0
; Total bytes of code: 38


```

Reply:
While mine is 4:
```csharp
    public static double RoundAWS_Egor(double value)
    {
        Vector128<double> valueVec = Vector128.CreateScalarUnsafe(value);
        Vector128<double> tmp = Sse2.And(valueVec, Vector128.Create(0x8000000000000000UL).AsDouble());
        tmp = Sse2.Or(tmp, Vector128.Create(0x3fdfffffffffffffUL).AsDouble());
        tmp = Sse2.Add(tmp, valueVec);
        return Sse41.RoundToZeroScalar(tmp).ToScalar();
    }
```
```asm
; Method Program:RoundAWS_Egor(double):double
G_M51304_IG01:
       vzeroupper 
						;; bbWeight=1    PerfScore 1.00

G_M51304_IG02:
       vandpd   xmm1, xmm0, xmmword ptr [reloc @RWD00]
       vorpd    xmm1, xmm1, xmmword ptr [reloc @RWD16]

       vaddpd   xmm0, xmm1, xmm0
       vroundsd xmm0, xmm0, xmm0, 3
						;; bbWeight=1    PerfScore 14.00

G_M51304_IG03:
       ret      
						;; bbWeight=1    PerfScore 1.00
RWD00  	dq	8000000000000000h, 8000000000000000h
RWD16  	dq	3FDFFFFFFFFFFFFFh, 3FDFFFFFFFFFFFFFh
; Total bytes of code: 30
```

Reply:
> Can't fit your impl into 4, it's 6 for me:

That looks to be duplicating some operations and is unnecessarily using an extra constant.
```csharp
        var val = Vector128.CreateScalarUnsafe(x);
        var negZero = Vector128.Create(0x7FFFFFFF_FFFFFFFF).AsDouble();
        
        var abs = Sse2.And(val, negZero);
        var res = Sse41.RoundToPositiveInfinityScalar(abs);
        
        return Sse2.Or(res, Sse2.AndNot(val, negZero)).ToScalar();
```

generates:
```asm
    L0003: vmovupd xmm0, [0x7ffb39740470]
    L000b: vandpd xmm2, xmm1, xmm0
    L000f: vroundsd xmm2, xmm2, xmm2, 2
    L0015: vandnpd xmm0, xmm1, xmm0
    L0019: vorpd xmm0, xmm2, xmm0
```

Even with the 1 extra instruction; it will likely be faster due to not needing a secondary memory read.

Reply:
My implementation would also solve any issues that would otherwise be caused by rounding or precision due to `BitDecrement(0.5)` not being exactly represented.

There may even be another trick here using `xor`, but it's likely not worth micro-optimizing the one extra instruction.

Reply:
A different ordering that generates the following would likely be even better due to pipelining:
```asm
    vmovupd xmm0, [0x7ffb39740470]
    vandpd xmm2, xmm1, xmm0
    vandnpd xmm0, xmm1, xmm0
    vroundsd xmm2, xmm2, xmm2, 2
    vorpd xmm0, xmm2, xmm0
```

This is because the `and` + `andnot` can be dispatched simultaneously and the `andnot` will have finished by the time `or` needs it.

Reply:
Ah, two bits...
1. I mixed up the operand ordering for `AndNot`
2. I forgot that `RoundToPositiveInfinity` isn't midpoint; its `Ceiling`

In either case, I'm still not convinced the `x + Math.BitDecrement(0.5)` is correct either. I'm fairly certain this is going to run into precision issues.

Reply:
Chatted with Tanner in Discord and gave up on Xarch impl